components/ipfs/ipfs_service.cc

```diff
@@ -272,6 +294,56 @@ std::unique_ptr<network::SimpleURLLoader> IpfsService::CreateURLLoader(
   return url_loader;
 }
 
+void IpfsService::ImportLinkToIpfs(const GURL& url,
+                                   ipfs::ImportCompletedCallback callback) {
+  ReentrancyCheck reentrancy_check(&reentrancy_guard_);
+  if (!IsDaemonLaunched()) {
+    StartDaemonAndLaunch(base::BindOnce(&IpfsService::ImportLinkToIpfs,
+                                        weak_factory_.GetWeakPtr(), url,
+                                        std::move(callback)));
+    return;
+  }
+  size_t key = base::FastHash(base::as_bytes(base::make_span(url.spec())));
+  if (importers_.count(key))
+    return;
+
+  auto import_completed_callback =
+      base::BindOnce(&IpfsService::OnImportFinished, weak_factory_.GetWeakPtr(),
+                     std::move(callback), key);
+  importers_[key] = std::make_unique<IpfsLinkImportWorker>(
+      context_, server_endpoint_, std::move(import_completed_callback), url);
+}
+
+void IpfsService::ImportTextToIpfs(const std::string& text,
+                                   const std::string& host,
+                                   ipfs::ImportCompletedCallback callback) {
+  ReentrancyCheck reentrancy_check(&reentrancy_guard_);
+  if (!IsDaemonLaunched()) {
+    StartDaemonAndLaunch(base::BindOnce(&IpfsService::ImportTextToIpfs,
+                                        weak_factory_.GetWeakPtr(), text, host,
+                                        std::move(callback)));
+    return;
+  }
+  size_t key = base::FastHash(base::as_bytes(base::make_span(text)));
+  if (importers_.count(key))
+    return;
+  auto import_completed_callback =
+      base::BindOnce(&IpfsService::OnImportFinished, weak_factory_.GetWeakPtr(),
+                     std::move(callback), key);
+  importers_[key] = std::make_unique<IpfsTextImportWorker>(
```

Comment:
if `IpfsTextImportWorker::StartImportText` early returns in case `text` or `host` is empty, it will synchronously call the callback, that won't erase the failed importer from the map (because it is not inserted yet), correct? After that it will be inserted and will live in the map forever. So maybe we'd better avoid running complex logic that involves chains of callbacks in constructors